rs/ethereum/cketh/minter/DepositHelperWithSubaccount.sol

```diff
@@ -1,8 +1,467 @@
+     * non-reverting calls are assumed to be successful. Meant to be used with tokens that require the approval
+     * to be set to zero before setting it to a non-zero value, such as USDT.
+     */
+    function forceApprove(IERC20 token, address spender, uint256 value) internal {
+        bytes memory approvalCall = abi.encodeCall(token.approve, (spender, value));
+
+        if (!_callOptionalReturnBool(token, approvalCall)) {
+            _callOptionalReturn(token, abi.encodeCall(token.approve, (spender, 0)));
+            _callOptionalReturn(token, approvalCall);
+        }
+    }
+
+    /**
+     * @dev Imitates a Solidity high-level call (i.e. a regular function call to a contract), relaxing the requirement
+     * on the return value: the return value is optional (but if data is returned, it must not be false).
+     * @param token The token targeted by the call.
+     * @param data The call data (encoded using abi.encode or one of its variants).
+     */
+    function _callOptionalReturn(IERC20 token, bytes memory data) private {
+        // We need to perform a low level call here, to bypass Solidity's return data size checking mechanism, since
+        // we're implementing it ourselves. We use {Address-functionCall} to perform this call, which verifies that
+        // the target address contains contract code and also asserts for success in the low-level call.
+
+        bytes memory returndata = address(token).functionCall(data);
+        if (returndata.length != 0 && !abi.decode(returndata, (bool))) {
+            revert SafeERC20FailedOperation(address(token));
+        }
+    }
+
+    /**
+     * @dev Imitates a Solidity high-level call (i.e. a regular function call to a contract), relaxing the requirement
+     * on the return value: the return value is optional (but if data is returned, it must not be false).
+     * @param token The token targeted by the call.
+     * @param data The call data (encoded using abi.encode or one of its variants).
+     *
+     * This is a variant of {_callOptionalReturn} that silents catches all reverts and returns a bool instead.
+     */
+    function _callOptionalReturnBool(IERC20 token, bytes memory data) private returns (bool) {
+        // We need to perform a low level call here, to bypass Solidity's return data size checking mechanism, since
+        // we're implementing it ourselves. We cannot use {Address-functionCall} here since this should return false
+        // and not revert is the subcall reverts.
+
+        (bool success, bytes memory returndata) = address(token).call(data);
+        return success && (returndata.length == 0 || abi.decode(returndata, (bool))) && address(token).code.length > 0;
+    }
+}
+
+// File: ck-deposit.sol
+
+
 
 pragma solidity ^0.8.20;
 
-import {SafeERC20, IERC20} from "https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v5.0.2/contracts/token/ERC20/utils/SafeERC20.sol";
```

Comment:
FYI: This is the exact same bytecode, just the version on the right was flatten using remix. The reason is that the import was not resolved by `foundry`. You can check that the version on the right matches the one used in production https://etherscan.io/address/0x18901044688D3756C35Ed2b36D93e6a5B8e00E68#code

Reply:
Just curious, how hard is it to use some other tool to resolve the imports at run time?

Reply:
Theoretically it should already be possible:

1. one can install the OpenZeppellin contracts with foundry (see [docs](url))
2. One can then use the `--remappings` option (see [forge create](https://book.getfoundry.sh/reference/forge/forge-create)) to remap the import, which according to the solidity compiler should also work with URLs (see [solidity docs](https://docs.soliditylang.org/en/latest/path-resolution.html#using-urls-in-imports))

However, when trying this locally I could not get it to work 🤷‍♂️ . An alternative instead of flattening would be to use relative paths, such as `@openzeppelin/contracts` instead of an absolute URL `https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v5.0.2/contracts`; however, I do like being able to pinpoint to the exact version used. The flattened version is also the one that was verified by Etherscan.

Reply:
I will close this thread to merge, but let me know if we should explore other options for this.